### ISSUE TITLE
System.Net.Http built on Ubuntu 16.04 cannot run on Ubuntu 18

### ISSUE THREAD

**Issue by wfurt** [dotnet/corefx#35838]
This is root cause for #34599 and symptoms are:

```
System.TypeInitializationException : The type initializer for 'System.Net.Http.CurlHandler' threw an exception.
---- System.TypeInitializationException : The type initializer for 'Http' threw an exception.
-------- System.TypeInitializationException : The type initializer for 'HttpInitializer' threw an exception.
------------ System.DllNotFoundException : Unable to load shared library 'System.Net.Http.Native' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libSystem.Net.Http.Native: cannot open shared object file: No such file or directory
```

This is because  System.Net.Http.Native has dependency on libcurl.so.4. And that is link to libcurl3 on build server. That would not still be too bad but Ubuntu uses symbol versioning. 
So even if we depend on libcurl4 we need libcurl3 symbols. 
On Ubuntu 18 we can still open the shared library but the symbols do not match: 

```
robox@DDARM64S-032:~/wfurt/dotnet-3.x$ ldd ./shared/Microsoft.NETCore.App/3.0.0-preview4-27506-6/System.Net.Http.Native.so| grep curl
./shared/Microsoft.NETCore.App/3.0.0-preview4-27506-6/System.Net.Http.Native.so: /usr/lib/aarch64-linux-gnu/libcurl.so.4: version `CURL_OPENSSL_3' not found (required by ./shared/Microsoft.NETCore.App/3.0.0-preview4-27506-6/System.Net.Http.Native.so)
	libcurl.so.4 => /usr/lib/aarch64-linux-gnu/libcurl.so.4 (0x0000ffffa4f89000)
```

For x64 we avoid this by building on Centos where symbol versioning is not used.

**Comment by jkotas** [comment]
@stephentoub Is there a good reason for keeping CurlHandler and dependency on libcurl around? Do we plan to delete it at some point?

**Comment by davidsh** [comment]
At some point in the future, yes, we will delete CurlHandler from master branch and the libcurl dependency.  But we are not yet at feature parity with SocketsHttpHandler.  We need to finish the HTTP/2 work.  So, removing CurlHandler is probably a post .NET Core 3.0 thing.

@karelz

**Comment by stephentoub** [comment]
We should at least think through what it would mean to delete it for .NET Core 3, given that it's a major version where we can be a little more flexible with such changes.  IIRC, the main things holding us back were HTTP/2 support and better proxy support, both of which should be addressed for 3.0.

Regardless, we had a goal that it would be completely optional, both off by default and that .NET Core could be built from source without the libcurl dependency.  At the very least we should strive to ensure that's the case for 3.0.

**Comment by wfurt** [comment]
note that this is currently problem only on ARM where we use Ubuntu based rootfs to cross-compile. (for X64 we build on Centos) Since there is no prior ARM64 this may not be consider as regression if we do not have curl.

Current plan how to fix this is to create Centos based rootfs for ARM and build same way as we do for X64.

**Comment by karelz** [comment]
arm64 build issue - can slip post ZBB (maybe even arm32)

**Comment by ItalyPaleAle** [comment]
Have this issue while running the RPi build of .NET Core (2.1.11) on Raspbian based on Debian 10 (which uses libicu4)

**Comment by wfurt** [comment]
This should not happen when using default HTTP handler. When somebody explicitly need CurlHttpHandler, the workaround is to install legacy libcurl3. That would normally remove curl binary as libcurl4 and libcurl3 are conflicting. The conflict can be solved with some packaging and LD_LIBRARY_PATH trickery or with containers. 
Risk of changing ARM build late in release is non-trivial. 
This should be probably fixed in 3.1

**Comment by karelz** [comment]
Triage: Will go away with removal of CurlHandler in 5.0.
